services/galley/src/Galley/Cassandra/MakeFeature.hs

```diff
@@ -0,0 +1,440 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | Abstraction to fetch and store feature values from and to the database.
+module Galley.Cassandra.MakeFeature where
+
+import Cassandra
+import Cassandra qualified as C
+import Data.Functor
+import Data.Functor.Identity
+import Data.Id
+import Data.Kind
+import Data.List.Singletons (Length)
+import Data.Misc (HttpsUrl)
+import Data.Singletons (demote)
+import Data.Time
+import GHC.TypeNats
+import Galley.Cassandra.FeatureTH
+import Galley.Cassandra.Instances ()
+import Generics.SOP
+import Imports hiding (Generic, Map)
+import Wire.API.Conversation.Protocol (ProtocolTag)
+import Wire.API.MLS.CipherSuite
+import Wire.API.Team.Feature
+
+-- | This is necessary in order to convert an @NP f xs@ type to something that
+-- CQL can understand.
+$(generateTupleP)
+
+class MakeFeature cfg where
+  type FeatureRow cfg :: [Type]
+  type FeatureRow cfg = '[FeatureStatus]
+
+  featureColumns :: NP (K String) (FeatureRow cfg)
+
+  mkFeature :: NP Maybe (FeatureRow cfg) -> DbFeature cfg
+  default mkFeature ::
+    (FeatureRow cfg ~ '[FeatureStatus]) =>
+    NP Maybe (FeatureRow cfg) ->
+    DbFeature cfg
+  mkFeature = foldMap dbFeatureStatus . hd
+
+  unmkFeature :: LockableFeature cfg -> NP Maybe (FeatureRow cfg)
+  default unmkFeature ::
+    (FeatureRow cfg ~ '[FeatureStatus]) =>
+    LockableFeature cfg ->
+    NP Maybe (FeatureRow cfg)
+  unmkFeature feat = Just feat.status :* Nil
```

Comment:
maybe these should be factored out and called `defaultUnmkFeature` and `defaultMkFeature` to avoid duplicating the default code for the locked versions of things

Reply:
I'm not sure how that would work. Features with a lock status as well as status cannot really share much code, since they have a different type, so need a different pattern match. The duplication is minimal anyway, so I think this is fine.